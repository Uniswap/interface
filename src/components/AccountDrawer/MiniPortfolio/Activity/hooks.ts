import { TransactionStatus, useActivityQuery } from 'graphql/data/__generated__/types-and-hooks'
import { useEffect, useMemo } from 'react'
import { usePendingOrders } from 'state/signatures/hooks'
import { usePendingTransactions, useTransactionCanceller } from 'state/transactions/hooks'
import { useFormatter } from 'utils/formatNumbers'

import { useLocalActivities } from './parseLocal'
import { parseRemoteActivities } from './parseRemote'
import { Activity, ActivityMap } from './types'

/** Detects transactions from same account with the same nonce and different hash */
function findCancelTx(localActivity: Activity, remoteMap: ActivityMap, account: string): string | undefined {
  // handles locally cached tx's that were stored before we started tracking nonces
  if (!localActivity.nonce || localActivity.status !== TransactionStatus.Pending) return undefined

  for (const remoteTx of Object.values(remoteMap)) {
    if (!remoteTx) continue

    // A pending tx is 'cancelled' when another tx with the same account & nonce but different hash makes it on chain
    if (
      remoteTx.nonce === localActivity.nonce &&
      remoteTx.from.toLowerCase() === account.toLowerCase() &&
      remoteTx.hash.toLowerCase() !== localActivity.hash.toLowerCase() &&
      remoteTx.chainId === localActivity.chainId
    ) {
      return remoteTx.hash
    }
  }

  return undefined
}

/** Deduplicates local and remote activities */
function combineActivities(localMap: ActivityMap = {}, remoteMap: ActivityMap = {}): Array<Activity> {
  const txHashes = [...new Set([...Object.keys(localMap), ...Object.keys(remoteMap)])]

  return txHashes.reduce((acc: Array<Activity>, hash) => {
    const localActivity = (localMap?.[hash] ?? {}) as Activity
    const remoteActivity = (remoteMap?.[hash] ?? {}) as Activity

    if (localActivity.cancelled) {
      // Hides misleading activities caused by cross-chain nonce collisions previously being incorrectly labelled as cancelled txs in redux
      if (localActivity.chainId !== remoteActivity.chainId) {
        acc.push(remoteActivity)
        return acc
      }
      // Remote data only contains data of the cancel tx, rather than the original tx, so we prefer local data here
      acc.push(localActivity)
    } else {
      // Generally prefer remote values to local value because i.e. remote swap amounts are on-chain rather than client-estimated
      acc.push({ ...localActivity, ...remoteActivity } as Activity)
    }

    return acc
  }, [])
}

export function useAllActivities(account: string) {
  const { formatNumberOrString } = useFormatter()
  const { data, loading, refetch } = useActivityQuery({
    variables: { account },
    errorPolicy: 'all',
    fetchPolicy: 'cache-first',
  })

  const localMap = useLocalActivities(account)
  const remoteMap = useMemo(
    () => parseRemoteActivities(formatNumberOrString, data?.portfolios?.[0].assetActivities),
    [data?.portfolios, formatNumberOrString]
  )
  const updateCancelledTx = useTransactionCanceller()

  /* Updates locally stored pendings tx's when remote data contains a conflicting cancellation tx */
  useEffect(() => {
    if (!remoteMap) return

    Object.values(localMap).forEach((localActivity) => {
      if (!localActivity) return

      const cancelHash = findCancelTx(localActivity, remoteMap, account)

      if (cancelHash) updateCancelledTx(localActivity.hash, localActivity.chainId, cancelHash)
    })
  }, [account, localMap, remoteMap, updateCancelledTx])

  const combinedActivities = useMemo(
    () => (remoteMap ? combineActivities(localMap, remoteMap) : undefined),
    [localMap, remoteMap]
  )

  return { loading, activities: combinedActivities, refetch }
}

export function usePendingActivity() {
  const pendingTransactions = usePendingTransactions()
  const pendingOrders = usePendingOrders()

  const hasPendingActivity = pendingTransactions.length > 0 || pendingOrders.length > 0
  const pendingActivityCount = pendingTransactions.length + pendingOrders.length

  return { hasPendingActivity, pendingActivityCount }
}
