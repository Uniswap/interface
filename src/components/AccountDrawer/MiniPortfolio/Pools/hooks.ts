import { Token } from '@uniswap/sdk-core'
import { AddressMap } from '@uniswap/smart-order-router'
import { abi as MulticallABI } from '@uniswap/v3-periphery/artifacts/contracts/lens/UniswapInterfaceMulticall.sol/UniswapInterfaceMulticall.json'
import { abi as NFTPositionManagerABI } from '@uniswap/v3-periphery/artifacts/contracts/NonfungiblePositionManager.sol/NonfungiblePositionManager.json'
import { useWeb3React } from '@web3-react/core'
import { MULTICALL_ADDRESS, NONFUNGIBLE_POSITION_MANAGER_ADDRESSES as V3NFT_ADDRESSES } from 'constants/addresses'
import { isSupportedChain, SupportedChainId } from 'constants/chains'
import { RPC_PROVIDERS } from 'constants/providers'
import { BaseContract } from 'ethers/lib/ethers'
import { ContractInput, useUniswapPricesQuery } from 'graphql/data/__generated__/types-and-hooks'
import { toContractInput } from 'graphql/data/util'
import useStablecoinPrice from 'hooks/useStablecoinPrice'
import { useMemo } from 'react'
import { NonfungiblePositionManager, UniswapInterfaceMulticall } from 'types/v3'
import { getContract } from 'utils'
import { CurrencyKey, currencyKey, currencyKeyFromGraphQL } from 'utils/currencyKey'

import { PositionInfo } from './cache'

type ContractMap<T extends BaseContract> = { [key: number]: T }

// Constructs a chain-to-contract map, using the wallet's provider when available
function useContractMultichain<T extends BaseContract>(
  addressMap: AddressMap,
  ABI: any,
  chainIds?: SupportedChainId[]
): ContractMap<T> {
  const { chainId: walletChainId, provider: walletProvider } = useWeb3React()

  return useMemo(() => {
    const relevantChains =
      chainIds ??
      Object.keys(addressMap)
        .map((chainId) => parseInt(chainId))
        .filter(isSupportedChain)

    return relevantChains.reduce((acc: ContractMap<T>, chainId) => {
      const provider = walletProvider && walletChainId === chainId ? walletProvider : RPC_PROVIDERS[chainId]
      acc[chainId] = getContract(addressMap[chainId], ABI, provider) as T
      return acc
    }, {})
  }, [ABI, addressMap, chainIds, walletChainId, walletProvider])
}

export function useV3ManagerContracts(chainIds: SupportedChainId[]): ContractMap<NonfungiblePositionManager> {
  return useContractMultichain<NonfungiblePositionManager>(V3NFT_ADDRESSES, NFTPositionManagerABI, chainIds)
}

export function useInterfaceMulticallContracts(chainIds: SupportedChainId[]): ContractMap<UniswapInterfaceMulticall> {
  return useContractMultichain<UniswapInterfaceMulticall>(MULTICALL_ADDRESS, MulticallABI, chainIds)
}

type PriceMap = { [key: CurrencyKey]: number | undefined }
export function usePoolPriceMap(positions: PositionInfo[] | undefined) {
  const contracts = useMemo(() => {
    if (!positions || !positions.length) return []
    // Avoids fetching duplicate tokens by placing in map
    const contractMap = positions.reduce((acc: { [key: string]: ContractInput }, { pool: { token0, token1 } }) => {
      acc[currencyKey(token0)] = toContractInput(token0)
      acc[currencyKey(token1)] = toContractInput(token1)
      return acc
    }, {})
    return Object.values(contractMap)
  }, [positions])

  const { data, loading } = useUniswapPricesQuery({ variables: { contracts }, skip: !contracts.length })

  const priceMap = useMemo(
    () =>
      data?.tokens?.reduce((acc: PriceMap, current) => {
        if (current) acc[currencyKeyFromGraphQL(current)] = current.project?.markets?.[0]?.price?.value
        return acc
      }, {}) ?? {},
    [data?.tokens]
  )

  return { priceMap, pricesLoading: loading && !data }
}

function useFeeValue(token: Token, fee: number | undefined, queriedPrice: number | undefined) {
  const stablecoinPrice = useStablecoinPrice(!queriedPrice ? token : undefined)
  return useMemo(() => {
    // Prefers gql price, as fetching stablecoinPrice will trigger multiple infura calls for each pool position
    const price = queriedPrice ?? (stablecoinPrice ? parseFloat(stablecoinPrice.toSignificant()) : undefined)
    const feeValue = fee && price ? fee * price : undefined

    return [price, feeValue]
  }, [fee, queriedPrice, stablecoinPrice])
}

export function useFeeValues(position: PositionInfo) {
  const [priceA, feeValueA] = useFeeValue(position.pool.token0, position.fees?.[0], position.prices?.[0])
  const [priceB, feeValueB] = useFeeValue(position.pool.token1, position.fees?.[1], position.prices?.[1])

  return { priceA, priceB, fees: (feeValueA || 0) + (feeValueB || 0) }
}
