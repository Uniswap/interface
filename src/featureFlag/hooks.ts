import { useFeatureFlagContext } from './provider'

enum Phase0Variant {
  Control = 'Control',
  Enabled = 'Enabled',
}

export function useFeatureFlagsLoaded(): boolean {
  return useFeatureFlagContext().isLoaded
}

export function usePhase0Flag(): Phase0Variant {
  const phase0Variant = useFeatureFlagContext().flags['phase0']
  switch (phase0Variant) {
    case 'enabled':
      return Phase0Variant.Enabled
    default:
      return Phase0Variant.Control
  }
}
