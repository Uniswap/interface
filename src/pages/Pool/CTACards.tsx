import { Trans } from '@lingui/macro'
import { useWeb3React } from '@web3-react/core'
import { AutoColumn } from 'components/Column'
import { getChainInfoOrDefault } from 'constants/chainInfo'
import styled from 'styled-components/macro'
import { ThemedText } from 'theme'

import { ExternalLink } from '../../theme'

const CTASection = styled.section`
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 8px;
  opacity: 0.8;

  ${({ theme }) => theme.deprecated_mediaWidth.deprecated_upToSmall`
    grid-template-columns: auto;
    grid-template-rows: auto;
  `};
`

const CTA1 = styled(ExternalLink)`
  padding: 16px;
  border-radius: 20px;
  display: flex;
  flex-direction: column;
  position: relative;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  border: 1px solid ${({ theme }) => theme.deprecated_bg3};

  * {
    color: ${({ theme }) => theme.textPrimary};
    text-decoration: none !important;
  }

  :hover {
    border: 1px solid ${({ theme }) => theme.deprecated_bg4};

    text-decoration: none;
    * {
      text-decoration: none !important;
    }
  }
`

const CTA2 = styled(ExternalLink)`
  position: relative;
  overflow: hidden;
  padding: 16px;
  border-radius: 20px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  border: 1px solid ${({ theme }) => theme.deprecated_bg3};

  * {
    color: ${({ theme }) => theme.textPrimary};
    text-decoration: none !important;
  }

  :hover {
    border: 1px solid ${({ theme }) => theme.deprecated_bg4};
    text-decoration: none !important;
    * {
      text-decoration: none !important;
    }
  }
`

const HeaderText = styled(ThemedText.DeprecatedLabel)`
  align-items: center;
  display: flex;

  font-weight: 400;
  font-size: 16px;
  ${({ theme }) => theme.deprecated_mediaWidth.deprecated_upToMedium`
    font-size: 16px;
  `};
`

const ResponsiveColumn = styled(AutoColumn)`
  grid-template-columns: 1fr;
  width: 100%;
  gap: 8px;

  ${({ theme }) => theme.deprecated_mediaWidth.deprecated_upToMedium`
    gap: 8px;
  `};
  justify-content: space-between;
`

export default function CTACards() {
  const { chainId } = useWeb3React()
  const { infoLink } = getChainInfoOrDefault(chainId)

  return (
    <CTASection>
      <CTA1 href="https://support.uniswap.org/hc/en-us/categories/8122334631437-Providing-Liquidity-">
        <ResponsiveColumn>
          <HeaderText>
            <Trans>Learn about providing liquidity</Trans> ↗
          </HeaderText>
          <ThemedText.DeprecatedBody fontWeight={400} style={{ alignItems: 'center', display: 'flex' }}>
            <Trans>Check out our v3 LP walkthrough and migration guides.</Trans>
          </ThemedText.DeprecatedBody>
        </ResponsiveColumn>
      </CTA1>
      <CTA2 data-testid="cta-infolink" href={infoLink + 'pools'}>
        <ResponsiveColumn>
          <HeaderText style={{ alignSelf: 'flex-start' }}>
            <Trans>Top pools</Trans> ↗
          </HeaderText>
          <ThemedText.DeprecatedBody fontWeight={400} style={{ alignSelf: 'flex-start' }}>
            <Trans>Explore Uniswap Analytics.</Trans>
          </ThemedText.DeprecatedBody>
        </ResponsiveColumn>
      </CTA2>
    </CTASection>
  )
}
