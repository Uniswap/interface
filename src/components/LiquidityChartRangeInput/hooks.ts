import { Currency } from '@uniswap/sdk-core'
import { FeeAmount } from '@uniswap/v3-sdk'
import { TickProcessed, usePoolActiveLiquidity } from 'hooks/usePoolTickData'
import { useCallback, useMemo } from 'react'

import { ChartEntry } from './types'

export function useDensityChartData({
  currencyA,
  currencyB,
  feeAmount,
}: {
  currencyA?: Currency
  currencyB?: Currency
  feeAmount?: FeeAmount
}) {
  const { isLoading, error, data } = usePoolActiveLiquidity(currencyA, currencyB, feeAmount)

  const formatData = useCallback(() => {
    if (!data?.length) {
      return undefined
    }

    const newData: ChartEntry[] = []

    for (let i = 0; i < data.length; i++) {
      const t: TickProcessed = data[i]

      const chartEntry = {
        activeLiquidity: parseFloat(t.liquidityActive.toString()),
        price0: parseFloat(t.price0),
      }

      if (chartEntry.activeLiquidity > 0) {
        newData.push(chartEntry)
      }
    }

    return newData
  }, [data])

  return useMemo(() => {
    return {
      isLoading,
      error,
      formattedData: !isLoading ? formatData() : undefined,
    }
  }, [isLoading, error, formatData])
}
