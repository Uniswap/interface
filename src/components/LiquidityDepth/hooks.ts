import { createContext, useEffect, useState } from 'react'
import { Currency } from '@uniswap/sdk-core'
import { FeeAmount } from '@uniswap/v3-sdk'
import JSBI from 'jsbi'
import { usePoolTickData } from 'hooks/usePoolTickData'
import { TickProcessed } from 'constants/ticks'

export interface ChartEntry {
  index: number
  isCurrent: boolean
  activeLiquidity: number
  price0: number
}

export const ChartContext = createContext<
  Partial<{ zoom: number; canZoomIn: boolean; canZoomOut: boolean; zoomIn: () => void; zoomOut: () => void }>
>({})

export function useDensityChartData({
  currencyA,
  currencyB,
  feeAmount,
}: {
  currencyA: Currency | undefined
  currencyB: Currency | undefined
  feeAmount: FeeAmount | undefined
}) {
  const [formattedData, setFormattedData] = useState<ChartEntry[] | undefined>()
  const [activeChartEntry, setActiveChartEntry] = useState<ChartEntry | undefined>()
  const [maxLiquidity, setMaxLiquidity] = useState<number>(0)

  const { loading, error, activeTick, tickData } = usePoolTickData(currencyA, currencyB, feeAmount)

  // clear data when inputs are cleared
  useEffect(() => {
    if (!currencyA || !currencyB || !feeAmount) {
      setFormattedData(undefined)
    }
  }, [currencyA, currencyB, feeAmount])

  useEffect(() => {
    function formatData() {
      if (!tickData?.length) {
        return
      }

      const newData: ChartEntry[] = []
      let maxLiquidity = JSBI.BigInt(0)

      for (let i = 0; i < tickData.length; i++) {
        const t: TickProcessed = tickData[i]
        const active = t.tickIdx === activeTick

        maxLiquidity = JSBI.greaterThan(tickData[i].liquidityActive, maxLiquidity)
          ? tickData[i].liquidityActive
          : maxLiquidity

        const chartEntry = {
          index: i,
          isCurrent: active,
          activeLiquidity: parseFloat(t.liquidityActive.toString()),
          price0: parseFloat(t.price0),
        }

        if (active) {
          setActiveChartEntry(chartEntry)
        }

        newData.push(chartEntry)
      }

      setMaxLiquidity(parseFloat(maxLiquidity.toString()))

      if (newData) {
        setFormattedData(newData)
      }
    }

    if (!loading) {
      formatData()
    }
  }, [loading, activeTick, tickData])

  return {
    loading,
    error,
    activeChartEntry,
    maxLiquidity,
    formattedData,
  }
}
