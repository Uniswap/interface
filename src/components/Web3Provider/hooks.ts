import { Web3ReactHooks } from '@web3-react/core'
import { Connector } from '@web3-react/types'
import { ConnectionType, gnosisSafe, injected, network } from 'connectors'
import { getConnectorForConnectionType, getHooksForConnectionType } from 'connectors/utils'
import { useEffect, useMemo } from 'react'
import { useAppSelector } from 'state/hooks'

import { isMobile } from '../../utils/userAgent'

export const BACKFILLABLE_WALLETS = [
  ConnectionType.COINBASE_WALLET,
  ConnectionType.WALLET_CONNECT,
  ConnectionType.INJECTED,
]
export const SELECTABLE_WALLETS = [...BACKFILLABLE_WALLETS, ConnectionType.FORTMATIC]

interface Connection {
  connector: Connector
  hooks: Web3ReactHooks
}

function getConnectionForConnectionType(connectionType: ConnectionType): Connection {
  return {
    connector: getConnectorForConnectionType(connectionType),
    hooks: getHooksForConnectionType(connectionType),
  }
}

export function useConnectors() {
  const selectedWallet = useAppSelector((state) => state.user.selectedWallet)
  return useMemo(() => {
    const connections: Connection[] = []

    // Always attempt to use to Gnosis Safe first, as we can't know if we're in a SafeContext.
    connections.push(getConnectionForConnectionType(ConnectionType.GNOSIS_SAFE))

    // Add the `selectedWallet` to the top so it's prioritized, then add the other selectable wallets.
    if (selectedWallet) {
      connections.push(getConnectionForConnectionType(selectedWallet))
    }
    connections.push(
      ...SELECTABLE_WALLETS.filter((wallet) => wallet !== selectedWallet).map(getConnectionForConnectionType)
    )

    // Add network connection last as it should be the fallback.
    connections.push(getConnectionForConnectionType(ConnectionType.NETWORK))

    // Convert to web3-react's representation of connectors.
    const web3Connectors: [Connector, Web3ReactHooks][] = connections.map(({ connector, hooks }) => [connector, hooks])
    return web3Connectors
  }, [selectedWallet])
}

async function connect(connector: Connector) {
  try {
    if (connector.connectEagerly) {
      await connector.connectEagerly()
    } else {
      await connector.activate()
    }
  } catch (error) {
    console.debug(`web3-react eager connection error: ${error}`)
  }
}

export function useEagerlyConnect() {
  const selectedWalletBackfilled = useAppSelector((state) => state.user.selectedWalletBackfilled)
  const selectedWallet = useAppSelector((state) => state.user.selectedWallet)

  const isMetaMask = !!window.ethereum?.isMetaMask

  useEffect(() => {
    connect(gnosisSafe)
    connect(network)

    if (isMobile && isMetaMask) {
      injected.activate()
    } else if (selectedWallet) {
      connect(getConnectorForConnectionType(selectedWallet))
    } else if (!selectedWalletBackfilled) {
      BACKFILLABLE_WALLETS.map(getConnectorForConnectionType).forEach(connect)
    }
    // The dependency list is empty so this is only run once on mount
  }, []) // eslint-disable-line react-hooks/exhaustive-deps
}
