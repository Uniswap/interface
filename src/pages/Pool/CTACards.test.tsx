import * as useV3Positions from 'hooks/useV3Positions'
import { render, screen } from 'test-utils/render'

import CTACards from './CTACards'

describe('CTAcard links', () => {
  it('renders mainnet link when chain is not supported', () => {
    jest.spyOn(useV3Positions, 'useV3Positions').mockImplementation(() => {
      return { loading: false, positions: undefined }
    })

    render(<CTACards />)
    expect(screen.getByTestId('cta-infolink')).toHaveAttribute('href', 'https://info.uniswap.org/#/pools')
  })
})
