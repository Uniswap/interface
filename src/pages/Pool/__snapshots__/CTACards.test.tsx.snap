// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTAcard links renders mainnet link when chain is not supported 1`] = `
<DocumentFragment>
  <section
    class="CTACards__CTASection-sc-1lbpl6-0 cIXeJs"
  >
    <a
      class="components__StyledLink-sc-88web0-5 kYFdcM CTACards__CTA1-sc-1lbpl6-1 eofYCX"
      href="https://help.uniswap.org/en/articles/5391541-providing-liquidity-on-uniswap-v3"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        class="Column__AutoColumn-sc-1r2yyln-2 CTACards__ResponsiveColumn-sc-1lbpl6-4 fYeWeg gjzWtJ"
      >
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL CTACards__HeaderText-sc-1lbpl6-3 kWiYgr css-kh5b9v"
        >
          Learn about providing liquidity ↗
        </div>
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL css-1frkxb8"
          style="align-items: center; display: flex;"
        >
          Check out our v3 LP walkthrough and migration guides.
        </div>
      </div>
    </a>
    <a
      class="components__StyledLink-sc-88web0-5 kYFdcM CTACards__CTA2-sc-1lbpl6-2 fbVxQY"
      href="https://info.uniswap.org/#/pools"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        class="Column__AutoColumn-sc-1r2yyln-2 CTACards__ResponsiveColumn-sc-1lbpl6-4 fYeWeg gjzWtJ"
      >
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL CTACards__HeaderText-sc-1lbpl6-3 kWiYgr css-kh5b9v"
          style="align-self: flex-start;"
        >
          Top pools ↗
        </div>
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL css-1frkxb8"
          style="align-self: flex-start;"
        >
          Explore Uniswap Analytics.
        </div>
      </div>
    </a>
  </section>
</DocumentFragment>
`;

exports[`CTAcard links renders specific link when chain is supported 1`] = `
<DocumentFragment>
  <section
    class="CTACards__CTASection-sc-1lbpl6-0 cIXeJs"
  >
    <a
      class="components__StyledLink-sc-88web0-5 kYFdcM CTACards__CTA1-sc-1lbpl6-1 eofYCX"
      href="https://help.uniswap.org/en/articles/5391541-providing-liquidity-on-uniswap-v3"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        class="Column__AutoColumn-sc-1r2yyln-2 CTACards__ResponsiveColumn-sc-1lbpl6-4 fYeWeg gjzWtJ"
      >
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL CTACards__HeaderText-sc-1lbpl6-3 kWiYgr css-kh5b9v"
        >
          Learn about providing liquidity ↗
        </div>
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL css-1frkxb8"
          style="align-items: center; display: flex;"
        >
          Check out our v3 LP walkthrough and migration guides.
        </div>
      </div>
    </a>
    <a
      class="components__StyledLink-sc-88web0-5 kYFdcM CTACards__CTA2-sc-1lbpl6-2 fbVxQY"
      href="https://info.uniswap.org/#/optimism/pools"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        class="Column__AutoColumn-sc-1r2yyln-2 CTACards__ResponsiveColumn-sc-1lbpl6-4 fYeWeg gjzWtJ"
      >
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL CTACards__HeaderText-sc-1lbpl6-3 kWiYgr css-kh5b9v"
          style="align-self: flex-start;"
        >
          Top pools ↗
        </div>
        <div
          class="theme__TextWrapper-sc-5lu8um-0 etkwRL css-1frkxb8"
          style="align-self: flex-start;"
        >
          Explore Uniswap Analytics.
        </div>
      </div>
    </a>
  </section>
</DocumentFragment>
`;
